src/jdk.incubator.code/share/classes/jdk/incubator/code/interpreter/Interpreter.java

```diff
@@ -487,12 +491,26 @@ static Object interpretOp(MethodHandles.Lookup l, OpContext oc, Op o) {
                     .asCollector(Object[].class, lo.parameters().size());
             Object fiInstance = MethodHandleProxies.asInterfaceInstance(fi, fProxy);
 
-            // If a quotable lambda proxy again to implement Quotable
+            // If a quotable lambda proxy again to add method Quoted quoted()
             if (Quotable.class.isAssignableFrom(fi)) {
-                return Proxy.newProxyInstance(l.lookupClass().getClassLoader(), new Class<?>[]{fi},
+                // Op.ofQuotable(Quotable q) expect q's class to have the method: Quoted quoted()
```

Comment:
This logic seems a bit expensive. I wonder if we could have a non-exported interface which extends Quoted and adds the quoted() method, which then the implementation knows about, but the outside world can't access. That would probably lead to cleaner code, here and elsewhere.

Reply:
I am not very concerned about performance in the Interpreter, but i am all for reducing the complexity!
Having an internal interface will reduce the complexity.

Reply:
> I am not very concerned about performance in the Interpreter, but i am all for reducing the complexity! Having an internal interface will reduce the complexity.

I think the internal interface will also improve performance of the new Op::ofQuotable (which currently uses reflection to discover the magic method it knows to be there)